src/tree.rs

```diff
@@ -3,79 +3,270 @@
-                        Kind::Folder => "Folder",
-                        Kind::Livemark => "Livemark",
-                        Kind::Separator => "Separator",
-                    })
+impl Tree {
+    /// Constructs a new rooted tree.
+    pub fn new(root: Item) -> Tree {
+        let mut index_by_guid = HashMap::new();
+        index_by_guid.insert(root.guid.to_string(), 0);
+        Tree { entries: vec![Entry { parent_index: None,
+                                     item: root,
+                                     level: 0,
+                                     child_indices: Vec::new(), }],
+               index_by_guid,
+               deleted_guids: HashSet::new(), }
+    }
+
+    pub fn deletions(&self) -> Vec<&str> {
+        self.deleted_guids
+            .iter()
+            .map(|guid| guid.as_ref())
+            .collect()
+    }
+
+    pub fn is_deleted(&self, guid: &str) -> bool {
+        self.deleted_guids.contains(guid)
+    }
+
+    pub fn note_deleted<T>(&mut self, guid: T)
+        where T: Into<String> {
+        self.deleted_guids.insert(guid.into());
+    }
+
+    pub fn guids(&self) -> Vec<&str> {
+        self.index_by_guid
+            .keys()
+            .map(|guid| guid.as_ref())
+            .chain(self.deleted_guids.iter().map(|guid| guid.as_ref()))
+            .collect()
+    }
+
+    pub fn node_for_guid<T>(&self, guid: T) -> Option<Node>
+        where T: AsRef<str> {
+        self.index_by_guid
+            .get(guid.as_ref())
+            .map(|index| self.node(*index))
+    }
+
+    pub fn insert<T>(&mut self, parent_guid: T, item: Item)
+        where T: AsRef<str> {
+        if self.index_by_guid.contains_key(&item.guid) {
+            panic!("Entry {} already exists in tree", &item.guid);
+        }
+        let child_index = self.entries.len();
+        let (parent_index, level) = match self.index_by_guid.get(parent_guid.as_ref()) {
+            Some(parent_index) => {
+                let parent = &mut self.entries[*parent_index];
+                if !parent.item.is_folder() {
+                    panic!("Can't insert {} into non-folder {}",
+                           &item.guid, &parent.item.guid);
+                }
+                parent.child_indices.push(child_index);
+                (*parent_index, parent.level + 1)
+            },
+            None => panic!("Missing parent {} for {}", &item.guid, parent_guid.as_ref()),
+        };
+        self.index_by_guid
+            .insert(item.guid.to_string(), child_index);
+        self.entries.push(Entry { parent_index: Some(parent_index),
+                                  item,
+                                  level,
+                                  child_indices: Vec::new(), });
+    }
+
+    fn node(&self, index: usize) -> Node {
+        Node(self, &self.entries[index])
     }
 }
 
-/// A complete, rooted tree with tombstones.
-pub struct Tree {
-    pub deleted_guids: HashSet<String>,
+impl PartialEq for Tree {
```

Comment:
See if you can just do this in the `mod test`, and if not, behind `#[cfg(test)]`